### ISSUE TITLE
Markdown: bad rendering of table with empty cells

### ISSUE THREAD

**Issue by baby-gnu** [mattermost/platform#6951]
#### Summary
On Mattermost 3.10.0, the tables with empty cells are not rendered properly.

#### Steps to reproduce

Write a message with the following table:

``` markdown
|       | Column1 | Column2 | Column 3 |
|--------------------------------------|
| line1 | l1c1    | l1c2    | l1c3     |
| line2 |         | l2c2    |          |
| line3 | l3c1    |         |          |
```

#### Expected behavior
We should have a 4x3 table perfectly aligned like with `multimarkdown`.

#### Observed behavior (that appears unintentional) 

The table is badly rendered like on the following screenshot.

![Rendered table](https://user-images.githubusercontent.com/1233212/28247631-afaa4f1a-6a34-11e7-89d1-12ee102e5766.png)

#### Possible fixes
No idea.

**Comment by hmhealey** [comment]
Hi @baby-gnu,

Thanks for the report. I must've broken it a few months ago when I was adding the ability to escape pipes inside a table. I've filed [a ticket](https://mattermost.atlassian.net/browse/PLT-7155) to go fix it for our next release in August.

One other thing, does that table render correctly in `multimarkdown`? I tried pasting it in Github, and it didn't render because you didn't include pipes to mark the columns in the second line. I'm asking because we're looking at redoing some of our Markdown code, and we were planning on matching Github's rendering, so the example you posted wouldn't work at all. Perhaps we could reconsider this case though if it's common.

**Comment by baby-gnu** [comment]
Hello @hmhealey,

Yes, it renders correctly with `multimarkdown`, no pipes on the headers separator line:

``` markdown
|       | Column1 | Column2 | Column3  |
|--------------------------------------|
| line1 | l1c1    | l1c2    | l1c3     |
| line2 |         | l2c2    |          |
| line3 | l3c1    |         |          |
```

is rendered as:

``` html
<table>
<col />
<thead>
<tr>
	<th> </th>
	<th>Column1</th>
	<th>Column2</th>
	<th>Column3</th>
</tr>
</thead>
<tbody>
<tr>
	<th>line1</th>
	<td>l1c1</td>
	<td>l1c2</td>
	<td>l1c3</td>
</tr>
<tr>
	<th>line2</th>
	<td> </td>
	<td>l2c2</td>
	<td> </td>
</tr>
<tr>
	<th>line3</th>
	<td>l3c1</td>
	<td> </td>
	<td> </td>
</tr>
</tbody>
</table>
```

With pipes to separate cells of the header separator line:

``` markdown
|       | Column1 | Column2 | Column3  |
|-------|---------|---------|----------|
| line1 | l1c1    | l1c2    | l1c3     |
| line2 |         | l2c2    |          |
| line3 | l3c1    |         |          |
```

is rendered as:

``` html
<table>
<col />
<col />
<col />
<col />
<thead>
<tr>
	<th> </th>
	<th>Column1</th>
	<th>Column2</th>
	<th>Column3</th>
</tr>
</thead>
<tbody>
<tr>
	<th>line1</th>
	<td>l1c1</td>
	<td>l1c2</td>
	<td>l1c3</td>
</tr>
<tr>
	<th>line2</th>
	<td> </td>
	<td>l2c2</td>
	<td> </td>
</tr>
<tr>
	<th>line3</th>
	<td>l3c1</td>
	<td> </td>
	<td> </td>
</tr>
</tbody>
</table>
```

I don't know what's the purpose of the `<col />` (1 time versus 4 times), but that's the only difference between both outputs.

Regards.

**Comment by hmhealey** [comment]
Cool, thanks. We'll have to consider if we can still support that when we redo the table rendering for our mobile app, but I think it should be possible

**Comment by baby-gnu** [comment]
There is no problem if Mattermost does not support the version without the pipes in the separator line, my old tables have them but are not rendered correctly anymore because of the empty cells :-/

Regards.

**Comment by lindy65** [comment]
A PR has been submitted which will fix the issue in the v4.1 release on August 16th.

**Comment by baby-gnu** [comment]
Thanks, I see it it #7015.

Regards.
